File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int maxHeight = height[0];
        int maxleft = 0;
        int index = 0;
        for(int i = 1; i < n; i++){
            if(maxHeight < height[i]){
                maxHeight = height[i];
                index = i;
            }
        }
        int water = 0;
        for(int i = 0; i < index; i++){
            if(maxleft > height[i]) water += maxleft - height[i];
            else maxleft = height[i];
        }
        int maxright = 0;
         for(int i = n-1; i > index; i--){
            if(maxright > height[i]) water += maxright - height[i];
            else maxright = height[i];
        }
        return water;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        n = len(height)
        maxHeight = height[0]
        maxleft = 0
        index = 0

        # Find the maximum height and its index
        for i in range(1, n):
            if maxHeight < height[i]:
                maxHeight = height[i]
                index = i

        water = 0

        # Calculate trapped water on the left side of the highest bar
        for i in range(index):
            if maxleft > height[i]:
                water += maxleft - height[i]
            else:
                maxleft = height[i]

        maxright = 0

        # Calculate trapped water on the right side of the highest bar
        for i in range(n-1, index, -1):
            if maxright > height[i]:
                water += maxright - height[i]
            else:
                maxright = height[i]

        return water